pkg/record/record.go

```diff
@@ -12,34 +12,121 @@
 	"github.com/conduitio/conduit/pkg/record/schema"
 )
 
-// Record ...
+const (
+	OperationCreate   Operation = iota + 1 // create
+	OperationUpdate                        // update
+	OperationDelete                        // delete
+	OperationSnapshot                      // snapshot
+)
+
+// Operation defines what triggered the creation of a record.
+type Operation int
+
+func (i Operation) MarshalText() ([]byte, error) {
+	return []byte(i.String()), nil
+}
+
+func (i *Operation) UnmarshalText(b []byte) error {
+	if len(b) == 0 {
+		return nil // empty string, do nothing
+	}
+
+	switch string(b) {
+	case OperationCreate.String():
+		*i = OperationCreate
+	case OperationUpdate.String():
+		*i = OperationUpdate
+	case OperationDelete.String():
+		*i = OperationDelete
+	case OperationSnapshot.String():
+		*i = OperationSnapshot
+	default:
+		// it's not a known operation, but we also allow Operation(int)
+		valIntRaw := strings.TrimSuffix(strings.TrimPrefix(string(b), "Operation("), ")")
+		valInt, err := strconv.Atoi(valIntRaw)
+		if err != nil {
+			return cerrors.Errorf("unknown operation %q", b)
+		}
+		*i = Operation(valInt)
+	}
+
+	return nil
+}
+
+// Record represents a single data record produced by a source and/or consumed
+// by a destination connector.
 type Record struct {
-	Position Position
-	Metadata map[string]string
-
-	// SourceID contains the source connector ID.
-	SourceID string
-
-	// CreatedAt represents the time when the change occurred in the source system.
-	// If that's impossible to find out, then it should be the time the change was detected by Conduit.
-	CreatedAt time.Time
-	// ReadAt represents the time at which Conduit read the record.
-	ReadAt time.Time
-
-	// Key and payload are guaranteed to be non-nil, always.
-	// However, they may be 'empty', i.e. not contain any real data.
-	Key     Data
-	Payload Data
+	// Position uniquely represents the record.
+	Position Position `json:"position"`
+	// Operation defines what triggered the creation of a record. There are four
+	// possibilities: create, update, delete or snapshot. The first three
+	// operations are encountered during normal CDC operation, while "snapshot"
+	// is meant to represent records during an initial load. Depending on the
+	// operation, the record will contain either the payload before the change,
+	// after the change, or both (see field Payload).
+	Operation Operation `json:"operation"`
+	// Metadata contains additional information regarding the record.
+	Metadata Metadata `json:"metadata"`
+
+	// Key represents a value that should identify the entity (e.g. database
+	// row).
+	Key Data `json:"key"`
+	// Payload holds the payload change (data before and after the operation
+	// occurred).
+	Payload Change `json:"payload"`
+}
+
+// Bytes returns the JSON encoding of the Record.
+func (r Record) Bytes() []byte {
+	if r.Metadata == nil {
+		// since we are dealing with a Record value this will not be seen
```

Comment:
If `Bytes()` is called often, that means, we might be increasing the memory usage?

Reply:
You mean because we are instantiating a map? Well sure, but only temporarily. The first question is if the map even escapes to the heap - if it stays on the stack then the memory will be reclaimed when the function exits. If on the other hand it lands on the heap it will be reclaimed next time the GC runs, because no object will reference the map anymore (the function receiver is a value so the `Metadata` field change is not visible outside of this function).

Reply:
I was asking because of the record as a whole, and mostly because of the payload. I commented on your comment above, since it mentions dealing with Record values.

Reply:
Maybe I should clarify - when I wrote "value" I meant a "Go value", as in "not a pointer". I'm saying that the receiver is a `Record` and not `*Record`.

I'm not sure I understand your concerns 🤔 I _assume_ you mean that because we use a value receiver the record will be copied when `Bytes` is called, which will increase the memory usage? I need to point out the copy is shallow, so if the record is carrying a huge payload it won't copy the payload - that's also the reason why we have to [delete the key](https://github.com/ConduitIO/conduit/blob/f4ce2e1007d55a4a8e99798ffcb3ec1a6979661a/pkg/record/record.go#L107) from metadata before returning from this function, because the metadata map is not copied.

Btw, maybe it's counterintuitive, but using values can be faster than using pointers, even though they are "copied" because Go does some optimizations in the background. I think this [blog post](https://www.ardanlabs.com/blog/2017/05/language-mechanics-on-stacks-and-pointers.html) explains it in depth.